### DISCUSSION TITLE
Script Widgets

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#82] [Roadmap]
## Progress
* High priority. 
* Probably a week or so away. 
* This is essentially how the "install window" worked if you can remember installing :D

## Feature explanation
Imagine creating a widget you could control from a script like this:

> _This is just the general idea_
```js
let channel = await widget(`<div>
<div class="timer">0</div>
<button class="button">Click me</button>
</div>`, {transparent:true, frame:false, width:200, height:200})

let count = 0
setInterval(()=> {
    channel.send(".timer", count++)
},1000=)

channel.on(".button$click", event => {
    console.log(event)
})
```

## Managing the process
1. The App would start the script and accept arguments 
2. Once it runs into a `widget` call, the App "detaches" the script process and launches the widget
3. The process would run until the widget closed

**Comment by zzeleznick** [comment]
I'd love this feature!

I regularly use [Memory Diag](https://apps.apple.com/us/app/memory-diag/id748212890?mt=12) to keep Chrome's greedy memory usage in check and envision being able to create a widget from a tray click action.

## Vision 

<img width="407" alt="memory_diag" src="https://user-images.githubusercontent.com/5779832/112763388-25278200-8fb9-11eb-99af-5b76e1a022df.png">


I know you've mentioned using Ink for terminal support and wonder if (in later versions) the widget could be a react component (e.g. [ink-counter-demo](https://apps.apple.com/us/app/memory-diag/id748212890?mt=12))?

**Reply by johnlindquist** [reply]
> React

Definitely. I'm thinking .kenv will eventually have a components dir. widget("foo.jsx") would load/compile based on the extension. Obviously there's a lot of work to put in before that's ready 🙂

**Comment by motleydev** [comment]
Watching :)

**Reply by johnlindquist** [reply]
They shipped in the last release :)

https://github.com/johnlindquist/kit/discussions/745
